serialize without id-arena fork

ydnar opened the pull request:
Updates `serde` support (serialization only) for `Resolve` and child types to work without a fork of the `id-arena` crate. It leans heavily on `#[serde(serialize_with = "...")]` to do the heavy lifting for foreign types.

The JSON serialization is opinionated:

1. Enums are flattened where possible, and variants are lowercased, to match values expressed in WIT versus the Rust implementation.
1. Function results are always expressed as a JSON array, with an optional `name` key. This is an affordance to parsers in languages with a less sophisticated type system (e.g. Go). `Anon` results will be a single-item JSON array, e.g.: `[{"type": <type>}]`
2. `ArenaId<T>` are serialized to JSON numbers as indexes into the arrays of items in the outer `Resolve` struct. The IDs are specific to this serialization, and are not globally unique.

ydnar committed:
wit-parser: add implementation of id_arena::ArenaBehavior for serde support

ydnar committed:
wit-parser: rename world arg to world_id

ydnar committed:
wit-parser: use InterfaceId and WorldId types

ydnar committed:
wit-parser: skip serialization of package name index

ydnar committed:
wit-parser: additional Serialize implementations

ydnar committed:
wit-parser: use derived Serialize for more types

ydnar committed:
wit-parser: change Params type from an alias to a single-value tuple with Deref/MutDeref

This allows us to implement Serialize on the Params type.

ydnar committed:
wit-parser: cleanup on aisle params

ydnar committed:
wit-parser: lowercase serialization for enums

Might change this?

ydnar committed:
wit-parser: implement Serialize in a function

This should allow wit-parser to serialize without the upstream patch to id_arena.

ydnar committed:
wit-parser: remove dependency on fork of id-arena crate

Mossaka commented on crates/wit-component/src/decoding.rs:
```diff
@@ -871,7 +871,7 @@ impl WitPackageDecoder<'_> {
             )
         } else {
-            Results::Named(
+            Results::Named(Params(
```
Could implement `Into` and `From` trait so don't need to wrap a `Params` outside but can call `into()`.

Mossaka commented on crates/wit-parser/Cargo.toml:
```diff
@@ -13,7 +13,7 @@ Tooling for parsing `*.wit` files and working with their contents.
 
 [dependencies]
-id-arena = { git = "https://github.com/Mossaka/id-arena-fork", rev = "cae2a0a63d505adec402289eea90c52d3205007a" }
+id-arena = "2"
```
nice!

Mossaka commented on crates/wit-parser/src/id_arena_.rs:
```diff
@@ -0,0 +1,97 @@
+    }
+}
+
+/// Our `ArenaBehavior` implementation.
```
Provide a better docs

Mossaka commented on crates/wit-parser/src/lib.rs:
```diff
@@ -562,7 +583,24 @@ pub struct Docs {
-pub type Params = Vec<(String, Type)>;
+// pub type Params = Vec<(String, Type)>;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
```
```suggestion
#[derive(Debug, Clone, PartialEq, Eq, Hash, Default)]
```

Mossaka commented on crates/wit-parser/src/lib.rs:
```diff
@@ -598,7 +636,7 @@ impl<'a> ExactSizeIterator for ResultsTypeIter<'a> {}
     // For the common case of an empty results list.
     pub fn empty() -> Results {
-        Results::Named(Vec::new())
+        Results::Named(Params(Vec::new()))
```
```suggestion
        Results::Named(Vec::new().into())
```

Mossaka commented on crates/wit-parser/src/live.rs:
```diff
@@ -25,8 +25,8 @@ impl LiveTypes {
 
-    pub fn add_world(&mut self, resolve: &Resolve, world: WorldId) {
-        let world = &resolve.worlds[world];
+    pub fn add_world(&mut self, resolve: &Resolve, world_id: WorldId) {
```
No need to change?

ydnar replied:
Probably not, maybe a future PR?

Mossaka reviewed (approved):
lgtm!

ydnar committed:
wit-parser: remove custom Serialize impl for Function

ydnar committed:
wit-parser: remove custom Serialize impl for Docs

ydnar committed:
wit-parser: custom Serialize implementation for Handle and TypeOwner

ydnar committed:
wit-parser: custom Serialize impl for WorldItem

ydnar committed:
wit-parser: Serialize enums with less code

ydnar committed:
wit-parser: implement lowercase derived impl Serialize for FunctionKind

ydnar committed:
wit-parser: serialize WIT types in JSON in their lowercase form

ydnar committed:
wit-component, wit-parser: remove custom Params type

Handle serialization using existing type alias to not break existing public API.

ydnar committed:
wit-parser: remove commented-out serialize fn

ydnar committed:
wit-parser: remove custom Serialize impl for Results

ydnar committed:
wit-parser: remove custom ArenaBehavior in favor of serialize_\* functions

ydnar committed:
wit-component: revert unnecessary line change

ydnar committed:
wit-parser: revert unrelated change

ydnar committed:
wit-parser: remove Serialize impl from Int

ydnar committed:
wit-parser: remove From<Int> for String

ydnar committed:
wit-parser: remove Serialize impl from FlagsRepr

Not needed?

Mossaka reviewed (approved):
LGTM! Thanks so much
